### PR TITLE
scripts: west: Add documentation for the multi-repo commands

### PR DESCRIPTION
**Pull request by ulfalizer**
Explain multi-repo concepts like manifests and projects, give an
overview of the currently implemented commands, and give an example of a
potential workflow.

There's no way to submit a multi-repo change for review yet, so this is
still experimental.

### CODE REVIEW

**Comment by ulfalizer** [comment]
CC @dbkinder @carlescufi @mbolivar

File: doc/west/repo-tool.rst [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+Command overview
+================
+
+This section gives a quick overview of the multi-repo commands, split up by
+functionality. All commands loosely mimic the corresponding Git command, but in
+a multi-repo context (``west fetch`` fetches upstream data without updating
+local work branches, etc.)
+
+Passing no projects to commands that accept a list of projects usually means to
+run the command for all projects.
```
Comment (dbkinder) [inline doc/west/repo-tool.rst]: maybe clarify with ``... for all projects listed in the manifest.`` ?

File: doc/west/repo-tool.rst [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+.. note::
+
+   For the most up-to-date documentation, see the command help texts (e.g.
```
Comment (dbkinder) [inline doc/west/repo-tool.rst]: We **could** pull the help content from the west python doc strings so it would automatically show up here... via http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html

**Review by dbkinder — DISMISSED** [review]
looks pretty good...

File: doc/west/index.rst [hunk 3]
```diff
@@ -25,6 +25,10 @@ pages.
 
    flash-debug.rst
 
+.. toctree::
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/index.rst]: Fixed

File: doc/west/repo-tool.rst [hunk 4]
```diff
@@ -0,0 +1,265 @@
+West: Multi-repository management
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+The rest of this page introduces multi-repo concepts and gives an overview of
+the multi-repo commands, along with an example workflow.
+
+.. note::
+
+   The multi-repo commands are meant to augment Git in minor ways for
+   multi-repo work, not replace it. For tasks that aren't multi-repo-related,
+   use plain Git commands.
+
+   This page tries to make it transparent what the multi-repo commands do
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst (hunk 1 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Yep, better

File: doc/west/repo-tool.rst [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+      $ west checkout -b xy-feature proj-x proj-y
+
+   This creates a Git branch named ``xy-feature`` in ``proj-x`` and ``proj-y``.
+   The new branches are automatically set up to track the ``manifest-rev``
+   branch in their respective repositories.
+
+   We work on the repositories as usual.
+
+3. While working on the feature, an urgent bug comes up that spans all
+   projects. We commit/stash the partially implemented feature and check out a
```
Comment (mbolivar) [inline doc/west/repo-tool.rst]: Hrm, west checkout doesn't actually stash the working tree; I think this could be cleared up a bit.

File: doc/west/repo-tool.rst [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+   ourselves, and then switch back:
+
+   .. code-block:: console
+
+      $ west checkout xy-feature
+
+   This command only affects the ``proj-x`` and ``proj-y`` repositories, since
+   only those have the ``xy-feature`` branch.
+
+5. We now update (rebase) the ``proj-1`` and ``proj-2`` repositories, to get
```
Comment (mbolivar) [inline doc/west/repo-tool.rst]: Looks like there's some leftovers from a previous version that used `proj-NUMBER` instead of `proj-LETTER` (elsewhere in the file too).

File: doc/west/repo-tool.rst [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+   upstream changes.
+
+   We want to update ``proj-x``. This could be done by running ``west pull
+   proj-x``, but since the ``xy-feature`` branch tracks the ``manifest-rev``
+   branch, we decide to mix it up a bit with a plain Git command inside the
+   ``proj-x`` repository:
+
+   .. code-block:: console
+
+      $ git pull
```
Comment (mbolivar) [inline doc/west/repo-tool.rst]: There's a missing `--rebase` here, no? Unlike west pull, git pull doesn't rebase by default.

**Review by mbolivar — CHANGES_REQUESTED** [review]
Thanks a lot for adding documentation! I really like the idea of "multi-repo" as a noun to describe what you get when you run west init + west clone.

Some random comments.

File: doc/west/repo-tool.rst [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+Running ``west init`` to initialize a West installation will clone the selected
+manifest repository. Running ``west init`` without specifying a manifest
+repository will use `a default manifest repository
+<https://github.com/zephyrproject-rtos/manifest>`_. A different manifest
+repository can be selected with ``west init --manifest-url URL``.
+
+The format of the West manifest is described by a `pykwalify
+<https://pypi.org/project/pykwalify/>`_ schema, found `here
+<https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/meta/west/commands/schema.yml>`_.
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+The ``manifest-rev`` branch
+***************************
+
+Each project managed through the multi-repo commands has a ``manifest-rev``
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Changed to a variant of the second suggestion.

File: doc/west/repo-tool.rst [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+branch that points to the project's revision as specified in the manifest. The
+``manifest-rev`` branch is updated whenever project data is fetched (see the
+`command overview`_ below).
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: The thinking was that the command overview explains when project data is fetched. I rephrased it a bit.

File: doc/west/repo-tool.rst [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+All branches created through the multi-repo commands track the ``manifest-rev``
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Hmm... I'd still want a "through" or the like in there, to make it clear that it's talking about branches created by the user through West.

Would you be fine with "all branches created through West"?

File: doc/west/repo-tool.rst [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+commands will clone/update projects.
+
+Except for ``west rebase``, all commands in this section implicitly update the
+manifest repository, fetch upstream data, and update the ``manifest-rev``
+branch.
+
+- ``west clone [-b BRANCH_NAME] [PROJECT ...]``: Clones the specified
+  projects (default: all projects).
+
+  An initial branch named after the project's revision is created in each
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Now explains how the name is derived in detail.

File: doc/west/repo-tool.rst [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+  cloned repository. A different name can be specified with ``-b``.
+
+  Like all branches created by the multi-repo commands, the initial branch
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: "Through West" maybe, to make it clear that it's talking about branches created by the user.

The talk about multi-repo commands is because we're bundling a bunch of other things in West as well, but it might be clear enough anyway.

File: doc/west/repo-tool.rst [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+
+     ``west clone`` is an alias for ``west fetch`` + ``west branch <name>``
+     (see below).
+
+- ``west pull [PROJECT ...]``: Fetches new upstream changes in each of the
+  specified projects (default: all projects) and rebases them to
+  ``manifest-rev``.
+
+  This corresponds to ``git pull --rebase``, with the upstream branch taken as
+  ``manifest-rev``.
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: It's both a local branch and an upstream branch. "Upstream" is more about the flow of data.

    git rev-parse --symbolic-full-name fix-foo@{upstream}
    refs/heads/manifest-rev

Maybe "local branch" could be added somewhere in there, but I wonder if it'd add confusion.

File: doc/west/repo-tool.rst [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+- ``west rebase [PROJECT ...]``: Rebases each of the specified projects
+  (default: all projects) to ``manifest-rev``.
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Changed

File: doc/west/repo-tool.rst [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+  projects are listed, along with the projects they appear in.
+
+- ``west checkout [-b] BRANCH_NAME [PROJECT ...]]``: Checks out the branch
+  ``BRANCH_NAME`` in each of the specified projects (default: all projects).
+  This command is a no-op for projects that do not have the specified branch.
+
+  With ``-b``, creates and checks out the branch if it does not exist in a
+  project (like ``git checkout -b``), instead of ignoring the project.
+
+Miscellanous commands
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+   projects:
+
+   .. code-block:: console
+
+      $ west init
+      $ west fetch
+
+   .. note::
+
+      ``west clone`` could be used instead of ``west fetch``, to automatically
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Was thinking of having `west clone` first, but thought this would be a nice and friendly way to explain detached HEADs and the difference between `fetch` and `clone`, which will probably come up elsewhere.

Any other opinions? Could change it if you think it'd be super confusing.

File: doc/west/repo-tool.rst (hunk 6 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: There was an implied "ourselves" in there. I rephrased it a bit. :)

File: doc/west/repo-tool.rst (hunk 7 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst (hunk 8 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Fixed

File: doc/west/repo-tool.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Changed it to "tracked branch". Maybe that's clearer.

File: doc/west/repo-tool.rst [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+A *manifest* is a `YAML <http://yaml.org/>`_ file that gives the URL and
+revision for each repository in the multi-repo (each *project*), possibly along
+with other information. Manifests are normally stored in a separate Git
+repository.
+
+Running ``west init`` to initialize a West installation will clone the selected
+manifest repository. Running ``west init`` without specifying a manifest
+repository will use `a default manifest repository
+<https://github.com/zephyrproject-rtos/manifest>`_ for Zephyr. A different
+manifest repository can be selected with ``west init --manifest-url URL``.
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: Let's use `-m` here, users are much more likely to use that

File: doc/west/repo-tool.rst [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+
+The format of the West manifest is described by a `pykwalify
+<https://pypi.org/project/pykwalify/>`_ schema, found `here
+<https://github.com/zephyrproject-rtos/west/blob/master/src/west/manifest-schema.yml>`_.
+
+The ``manifest-rev`` branch
+***************************
+
+West creates a branch named ``manifest-rev`` in each project, pointing to the
+commit at the project's revision, as specified in the manifest. The
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: `master` is not a commit. Maybe we need to rephrase this to express tracking branches?

File: doc/west/repo-tool.rst [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+``manifest-rev`` branch is updated whenever project data is fetched (the
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: Again, not quite right for manifests using SHAs or tags?

File: doc/west/repo-tool.rst (hunk 2 above)
Comment (carlescufi) [reply doc/west/repo-tool.rst]: Agreed, let's have the simple overview here

File: doc/west/repo-tool.rst [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+
+.. note::
+
+   For the most up-to-date documentation, see the command help texts (e.g.
+   ``west fetch --help``). Only the most important flags are mentioned here.
+
+Cloning and updating projects
+*****************************
+
+After running ``west init`` to initialize West (e.g., with the default Zephyr
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: Do we need the "(e.g....)" ?

File: doc/west/repo-tool.rst [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+manifest), the following commands will clone/update projects.
+
+Except for ``west rebase``, all commands in this section implicitly update the
+manifest repository, fetch upstream data, and update the ``manifest-rev``
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: "update the manifest and west repositories". I think we should make it clear that `west` updates both by default.

File: doc/west/repo-tool.rst [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+the multi-repo commands, along with an example workflow.
+
+.. note::
+
+   The multi-repo commands are meant to augment Git in minor ways for
+   multi-repo work, not replace it. For tasks that aren't multi-repo-related,
+   use plain Git commands.
+
+   This page explains what the West multi-repo commands do behind the scenes.
+
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: I'm missing a paragraph here explaining how `west` and `manifest` are special repositories also handled by `west`, and how that works with the rest. It's sort of connected to the multi-repo itself. Maybe what you've added below is enough, but I am not sure, especially for users that are less savvy.

File: doc/west/repo-tool.rst [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+  For qualified refs like ``refs/heads/foo``, the last component (``foo``) is
+  used. For SHA revisions, the generic name ``work`` is used.
+
+  A different branch name can be specified with ``-b``.
+
+  Like all branches created through West, the initial branch tracks
+  ``manifest-rev``.
+
+- ``west fetch [PROJECT ...]``: Fetches new upstream changes in each of the
+  specified projects (default: all projects), cloning them first if necessary.
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: Maybe mention that this is like `git fetch`, i.e. does not modify local branches, just `<remote>/*` branches.

File: doc/west/repo-tool.rst [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+
+  Extra arguments are passed as-is to ``git diff``.
+
+- ``west status [PROJECT ...] [ARGUMENT ...]``: Like ``west diff``, for
+  running ``git status``.
+
+- ``west update [--update-west] [--update-manifest]``: Updates the west and
+  manifest repositories.
+
+  Normally, the west and manifest repositories are updated automatically
```
Comment (carlescufi) [inline doc/west/repo-tool.rst]: Maybe have this sentence at the beginning of this page, as I mentioned earlier.

File: doc/west/repo-tool.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/west/repo-tool.rst]: I do believe that we should have both, but the default should use `west clone` because most people will use this as a reference. So I would reverse the main content and the note.

**Review by carlescufi — CHANGES_REQUESTED** [review]
Thanks Ulf, this is great stuff. Some comments added

File: doc/west/repo-tool.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/west/repo-tool.rst]: "All branches created using `west`"...

File: doc/west/repo-tool.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Alright, will go for that.

File: doc/west/repo-tool.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Ah, yeah, 'using' might be even better. I'll go for that.

File: doc/west/repo-tool.rst (hunk 19 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Changed, though I wonder if it's nicer to spell things out for seldomly-used flags like this.

File: doc/west/repo-tool.rst (hunk 22 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Used to say "after running ``west init`` to initialize Zephyr". That part was added as a compromise.

I could go back to the original version too, though I think it would have confused me.

File: doc/west/repo-tool.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Changed

File: doc/west/repo-tool.rst (hunk 20 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: `manifest-rev` is not a symbolic ref (like `HEAD` -> `refs/heads/master`, when you're not in detached `HEAD` mode), so it does point to a commit (whatever commit the revision resolves to).

I'll see if I can rephrase it a bit.

Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Rephrased it a bit. Is that better?

File: doc/west/repo-tool.rst (hunk 25 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: Now mentions that local branches aren't updated (except for `manifest-rev`).

**Comment by ulfalizer** [comment]
Updated the command docs to mention which commands default to all cloned projects, as opposed to all projects.

File: doc/west/repo-tool.rst (hunk 24 above)
Comment (ulfalizer) [reply doc/west/repo-tool.rst]: I added a note that mentions how the self-update for West itself is implemented ("To implement self-updates, west init also clones...").

IMO, the manifest update is the most important thing. The West self-update is more of an internal detail.

**Comment by ulfalizer** [comment]
Added some documentation for  `west forall` as well.

**Review by carlescufi — APPROVED** [review]
This is good enough for me as it is. We will be augmenting this as time goes by.

**Comment by carlescufi** [comment]
@ulfalizer please fix CI

**Comment by ulfalizer** [comment]
Fixed. Unrelated brokenness.

**Comment by carlescufi** [comment]
@mbolivar mind taking another look?

**Review by mbolivar — APPROVED** [review]
